Fixed and optimized Decimal.GetHashCode

pentp opened the pull request:
This fixes https://github.com/dotnet/coreclr/issues/2944 and contributes to https://github.com/dotnet/coreclr/issues/18249.
I'll add tests to CoreFX that break the current imprecise implementation.
I'm not sure if `HashCode.Combine` should be used here instead of a simple xor - it would be twice as slow.
And this also raises a question about decimal to double conversion - ideally, it should be normalized before conversion, but from a compatibility perspective it's not that clear.

_Updated benchmark with different input data sets:_
#### x64 (mostly normalized)
|  Method |     Mean |     Error |    StdDev | Scaled |
|-------- |---------:|----------:|----------:|-------:|
|  Native | 5.365 ns | 0.0127 ns | 0.0033 ns |   1.00 |
| CoreRT2 | 2.706 ns | 0.0103 ns | 0.0037 ns |   0.51 |

#### x64 (denormalized)
|  Method |     Mean |     Error |    StdDev | Scaled |
|-------- |---------:|----------:|----------:|-------:|
|  Native | 5.575 ns | 0.0275 ns | 0.0098 ns |   1.00 |
| CoreRT2 | 4.750 ns | 0.0243 ns | 0.0087 ns |   0.85 |

#### x64 (heavily denormalized)
|  Method |     Mean |     Error |    StdDev | Scaled |
|-------- |---------:|----------:|----------:|-------:|
|  Native | 6.162 ns | 0.0159 ns | 0.0057 ns |   1.00 |
| CoreRT2 | 6.482 ns | 0.0128 ns | 0.0046 ns |   1.05 |

#### x86 (mostly normalized)
|  Method |      Mean |     Error |    StdDev | Scaled |
|-------- |----------:|----------:|----------:|-------:|
|  Native |  7.706 ns | 0.1061 ns | 0.0378 ns |   1.00 |
| CoreRT2 |  4.093 ns | 0.0345 ns | 0.0123 ns |   0.54 |

#### x86 (denormalized)
|  Method |      Mean |     Error |    StdDev | Scaled |
|-------- |----------:|----------:|----------:|-------:|
|  Native |  7.870 ns | 0.0355 ns | 0.0127 ns |   1.00 |
| CoreRT2 | 14.292 ns | 0.0621 ns | 0.0222 ns |   1.82 |

#### x86 (heavily denormalized)
|  Method |      Mean |     Error |    StdDev | Scaled |
|-------- |----------:|----------:|----------:|-------:|
|  Native |  8.408 ns | 0.0426 ns | 0.0152 ns |   1.00 |
| CoreRT2 | 26.775 ns | 0.1536 ns | 0.0548 ns |   3.18 |

## x86 DIV
|  Method |     Mean |     Error |    StdDev | Scaled |
|-------- |---------:|----------:|----------:|-------:|
|  Native | 116.6 ns | 0.1997 ns | 0.0712 ns |   1.00 |
|  CoreRT | 235.5 ns | 1.0579 ns | 0.3773 ns |   2.02 |
| CoreRT2 | 113.9 ns | 0.2593 ns | 0.0674 ns |   0.98 |

morganbr commented:
If ```HashCode.Combine``` is only twice as slow as xor, it's probably worth doing since this xor would likely produce lots of collisions. I don't have a good sense of how often a Decimal is used as a Dictionary key though, which would dictate how important that is.

jkotas commented:
I would work better to submit functional improvements / bug fixes in decimal to CoreCLR first, or at the same time as they are submitted to CoreRT. It will make testing and keeping track of what it fixed where easier.

Would it be possible to add this (together with what it depends on to make it compile) to `Decimal.DecCalc.cs` in CoreCLR?

jkotas commented:
We do not randomize hashcodes of any built-in numeric types today. I would look at what to do about that separately from this issue. I do not think it makes sense to randomize for some and not for others.

pentp commented:
I think I can refactor it as a self-contained function for CoreCLR without bringing all of `DecCalc` along.

slavanap commented:
@pentp What input data do you use for performance tests? And how you run them?
For instance, for decimals ending with 23 zeros with scale = 23 there will be 5 invocations of Div96ByConst1XX functions, each of which contains 2 ulong divisions and 2 multiplications (1 for ulong and 1 for uint), that is more operations than in original CoreCLR implementation.

pentp commented:
@slavanap I've been using the same approach as in #4452, which is somewhat biased toward normalized decimals. I'll try measuring with a different set of decimals from https://github.com/dotnet/corefx/pull/30116 and see if the perf is unacceptable.
All the divisions are by a constant so actually multiplications by the reciprocal.

pentp commented:
With heavily denormalized decimals perf is following:
### x64
|  Method |     Mean |     Error |    StdDev | Scaled |
|-------- |---------:|----------:|----------:|-------:|
|  Native | 6.162 ns | 0.0159 ns | 0.0057 ns |   1.00 |
| CoreRT2 | 6.482 ns | 0.0128 ns | 0.0046 ns |   1.05 |
### x86
|  Method |      Mean |     Error |    StdDev | Scaled |
|-------- |----------:|----------:|----------:|-------:|
|  Native |  8.408 ns | 0.0426 ns | 0.0152 ns |   1.00 |
| CoreRT2 | 29.966 ns | 0.1981 ns | 0.0515 ns |   3.56 |

It looks like 32-bit RyuJIT doesn't use reciprocal multiplication for long divisions 😞

jkotas commented:
> It looks like 32-bit RyuJIT doesn't use reciprocal multiplication for long divisions

Are you testing this on .NET Core 2.1?

pentp commented:
Yes, latest coreclr source with debug JIT (for JitDump) with a release build:

```
if((low & 1) == 0 && iScale >= 1)
{
	const uint pow = 10;
	ulong div64 = high64 / pow;
	uint div = (uint)((((high64 - (uint)div64 * pow) << 32) | low) / pow);
	if(low == div * pow)
	{
		high64 = div64;
		low = div;
		iScale--;
	}
}
```
```
test     byte  ptr [ebp-18H], 1
jne      G_M65130_IG13
mov      dword ptr [ebp-14H], eax
cmp      eax, 1
jb       G_M65130_IG16
push     edi
push     esi
push     0
push     10
call     CORINFO_HELP_ULDIV
mov      dword ptr [ebp-80H], eax
mov      dword ptr [ebp-7CH], edx
mov      eax, dword ptr [ebp-80H]
mov      edx, dword ptr [ebp-7CH]
mov      dword ptr [ebp-48H], edx
mov      dword ptr [ebp-44H], eax
lea      ecx, [eax+4*eax]
shl      ecx, 1
mov      edx, esi
sub      edx, ecx
mov      ecx, dword ptr [ebp-18H]
mov      eax, ecx
or       eax, 0
or       edx, 0
push     edx
push     eax
push     0
push     10
call     CORINFO_HELP_ULDIV
mov      dword ptr [ebp-88H], eax
mov      dword ptr [ebp-84H], edx
mov      eax, dword ptr [ebp-88H]
mov      edx, dword ptr [ebp-84H]
lea      edx, [eax+4*eax]
shl      edx, 1
mov      ecx, dword ptr [ebp-18H]
cmp      ecx, edx
mov      dword ptr [ebp-18H], ecx
jne      SHORT G_M65130_IG15
mov      esi, dword ptr [ebp-44H]
mov      edi, dword ptr [ebp-48H]
mov      ecx, eax
mov      eax, dword ptr [ebp-14H]
dec      eax
mov      dword ptr [ebp-18H], ecx
```

There are some weird `xor eax, 0` lines also...

jkotas commented:
Ah ok. I does not do it for 64-bit division. It does it for 32-bit division only. I am not sure whether the magic division would be profitable for 64-bit on x86. The C++ compiler does not use it in this case either.

pentp commented:
I added a shortcut for 32-bit divisions that reduces `CORINFO_HELP_ULDIV` calls.
Refactored `VarDecDiv` to also use the new unscaling code and this brought x86 DIV from 1.03 to 0.98 on the benchmark.
With denormalized but still realistic numbers, x86 `GetHashCode` is 1.82x slower than currently - this feels like an acceptable trade-off considering that the current implementation is buggy.
